### PR TITLE
Adding "pod cache" command

### PR DESCRIPTION
**Pull request by AliSoftware**
Implementation of #3508

* [x] Implementation
* [x] Move cache listing (`cache_info_per_pod `) into the cache itself
* [x] Specs
* [x] CHANGELOG

### CODE REVIEW

**Comment by segiddins** [comment]
Please move the subcommands into their own files

File: lib/cocoapods/command/cache.rb [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+      # @return [Hash<String, Hash<Symbol, String>>]
+      #         A hash whose keys are the pod spec name
+      #         And values are a hash with the following keys:
+      #         :spec_file : path to the spec file
+      #         :name      : name of the pod
+      #         :version   : pod version
+      #         :release   : boolean to tell if that's a release pod
+      #         :slug      : the slug path where the pod cache is located
+      #
+      def cache_info_per_pod
```
Comment (segiddins) [inline lib/cocoapods/command/cache.rb]: please try and move some of this listing behavior into the cache itself?

File: lib/cocoapods/command/cache.rb [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        # Removes the specified cache
+        #
+        # @param [Array<Hash>] cache_infos
+        #        An array of caches to remove, each specified with the same
+        #        hash as cache_info_per_pod especially :spec_file and :slug
+        #
+        def remove_caches(cache_infos)
+          cache_infos.each do |info|
+            UI.message("Removing spec #{info[:spec_file]} (v#{info[:version]})")
+            FileUtils.rm(info[:spec_file])
```
Comment (segiddins) [inline lib/cocoapods/command/cache.rb]: the `rm` should be wrapped in a block argument to the `UI.message1

File: lib/cocoapods/command/cache/list.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+              print_pod_cache_infos(@pod_name, cache_descriptors)
+            end
+          end
+        end
+
+        private
+
+        # Prints the list of specs & pod cache dirs for a single pod name.
+        #
+        # This output is valid YAML so it can be parsed with 3rd party tools
```
Comment (segiddins) [inline lib/cocoapods/command/cache/list.rb]: make sure to escape the values

**Comment by segiddins** [comment]
@AliSoftware ping on specs?

**Comment by AliSoftware** [comment]
Oh right, thx for the reminder.

How would you suggest to spec that? Have a fixture that contains a directory structure representing some cache? (That's the way I would go, but I'm wondering if it's a strong enough test, given that if the cache structure changes, we won't catch it)

**Comment by segiddins** [comment]
I'd just say build up a temporary directory in the spec

**Comment by AliSoftware** [comment]
@segiddins done, should be ready to merge now :)

**Comment by segiddins** [comment]
Could you please put the cache fixture in a tarball and scrub it from history, so the repo doesn't get too bloated?

**Comment by AliSoftware** [comment]
fixed in b1228e70db763fa15d9824d4056abf31fff0a145

File: lib/cocoapods/command/cache/clean.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        end
+
+        def initialize(argv)
+          @pod_name = argv.shift_argument
+          @wipe_all = argv.flag?('all')
+          super
+        end
+
+        def run
+          if @pod_name.nil? # && @wipe_all
```
Comment (segiddins) [inline lib/cocoapods/command/cache/clean.rb]: please remove commented code

File: lib/cocoapods/downloader/cache.rb [hunk 5]
```diff
@@ -38,6 +38,37 @@ def download_pod(request)
...
+      #
+      def cache_descriptors_per_pod
+        specs_dir = root + 'Specs'
+        release_specs_dir = specs_dir + 'Release'
+        return {} unless specs_dir.exist?
+
+        spec_paths = specs_dir.find.select { |f| f.fnmatch('*.podspec.json') }
+        spec_paths.reduce({}) do |hash, spec_path|
+          spec = Specification.from_file(spec_path)
+          hash[spec.name] = [] if hash[spec.name].nil?
```
Comment (segiddins) [inline lib/cocoapods/downloader/cache.rb]: `||=`

**Comment by segiddins** [comment]
:fireworks: thanks @AliSoftware !

**Pull request merged**